cf-agent/package_module.c

```diff
@@ -0,0 +1,1604 @@
+{
+    for (const Rlist *rp = data; rp != NULL; rp = rp->next)
+    {
+        char *line = RlistScalarValue(rp);
+                   
+        if (StringStartsWith(line, "Error="))
+        {
+            Log(LOG_LEVEL_VERBOSE, "Caught error: %s", line + strlen("Error="));
+        }
+        else if (StringStartsWith(line, "ErrorMessage="))
+        {
+            Log(LOG_LEVEL_VERBOSE, "Caught error message: %s", 
+                line + strlen("ErrorMessage="));
+        }
+        else
+        {
+            Log(LOG_LEVEL_VERBOSE, "Caught unsupported error info: %s", line);
+        }
+    }
+}
+
+static int IsReadWriteReady(const IOData *io, int timeout_sec)
+{
+    fd_set  rset;
+    FD_ZERO(&rset);
+    FD_SET(io->read_fd, &rset);
+
+    struct timeval tv = {
+        .tv_sec = timeout_sec,
+        .tv_usec = 0,
+    };
+
+    int ret = select(io->read_fd + 1, &rset, NULL, NULL, &tv);
+
+    if(ret < 0)
+    {
+        Log(LOG_LEVEL_VERBOSE, "Failed checking for data. (select: %s)",
+            GetErrorStr());
+        return -1;
+    }
+
+    else if (FD_ISSET(io->read_fd, &rset))
+    {
+        return io->read_fd;
+    }
+
+    /* We have reached timeout */
+    if(ret == 0)
+    {
+        Log(LOG_LEVEL_DEBUG, "Timeout reading from package module.");
+        return 0;
+    }
+
+    Log(LOG_LEVEL_VERBOSE,
+        "Unknown outcome (ret > 0 but our only fds is not set). (select: %s)",
+        GetErrorStr());
+
+    return -1;
+}
+
+static Rlist *ReadDataFromPackageScript(const IOData *io)
+{
+    char buff[CF_BUFSIZE] = {0};
+    int red_data_size = 0;
+    
+    Buffer *data = BufferNew();
+    if (!data)
+    {
+        Log(LOG_LEVEL_VERBOSE, 
+            "Unable to allocate buffer for handling package module responses.");
+        return NULL;
+    }
+    
+    int timeout_seconds_left = PACKAGE_PROMISE_SCRIPT_TIMEOUT_SEC;
+    
+    while(!IsPendingTermination() && timeout_seconds_left > 0)
+    {
+        int fd = IsReadWriteReady(io, PACKAGE_PROMISE_TERMINATION_CHECK_SEC);
+        
+        if (fd < 0)
+        {
+            Log(LOG_LEVEL_VERBOSE, 
+                "Error reading data from package module: %s",
+                GetErrorStr());
+            return NULL;
+        }
+        else if (fd == io->read_fd)
+        {
+            ssize_t res = read(fd, buff, sizeof(buff) - 1);
+            if (res == -1)
+            {
+                if (errno == EINTR)
+                {
+                    continue;
+                }
+                else
+                {
+                    Log(LOG_LEVEL_VERBOSE,
+                        "Unable to read output from package module: %s",
+                        GetErrorStr());
+                    BufferDestroy(data);
+                    return NULL;
+                }
+            }
+            else if (res == 0) /* reached EOF */
+            {
+                break;
+            }
+            Log(LOG_LEVEL_DEBUG, "Data red from package module: %zu [%s]",
```

Comment:
In fact "red" occurs frequently. Although it's pronounced like "red", it's spelled "read" (both present and past has the same spelling, kudos to English for being so unambiguous....)